ext/soap/soap.c

```diff
@@ -1540,44 +1546,52 @@ PHP_METHOD(SoapServer, handle)
 				}
 			}
 #endif
-			if (zend_hash_find_ptr_lc(function_table, Z_STR(h->function_name)) != NULL ||
-			    ((service->type == SOAP_CLASS || service->type == SOAP_OBJECT) &&
-			     zend_hash_str_exists(function_table, ZEND_CALL_FUNC_NAME, sizeof(ZEND_CALL_FUNC_NAME)-1))) {
-				call_status = call_user_function(NULL, soap_obj, &h->function_name, &h->retval, h->num_params, h->parameters);
-				if (call_status != SUCCESS) {
-					php_error_docref(NULL, E_WARNING, "Function '%s' call failed", Z_STRVAL(h->function_name));
-					return;
+			zend_function *header_fn = zend_hash_find_ptr_lc(function_table, Z_STR(h->function_name));
+			if (UNEXPECTED(header_fn == NULL)) {
+				if (soap_obj_ce && soap_obj_ce->__call) {
+					header_fn = zend_get_call_trampoline_func(soap_obj_ce, Z_STR(function_name), false);
```

Comment:
When using `zend_get_call_trampoline_func` you should also release the trampoline function when you're done with it. This usually doesn't cause problems because as long as you're outside of a trampoline call, this function will store the trampoline in `EG(trampoline)`. It should be possible to test this by making a soap call from within a trampoline.

Reply:
You'll instantiate a SoapServer class and use `setObject` with a class that has a `__call` implementation.
The SoapServer::handle() function should be called from within another class's `__call` implementation.
That way, there are 2 trampolines active, and so the one invoked in this C code will use a heap-allocated function structure.

Reply:
I'm working on something more long term to see if I can get rid of the `function_name` zval of the FCI so manual trampoline handling is something that is a bit annoying, I agree.

Not sure if there should be a new API with a "backing" zval.